### ISSUE TITLE
Creating abstract ops for batch_normalization_train and batch_normalization_test

### ISSUE THREAD

**Issue by gvtulder** [Theano/Theano#5187]
This is a suggestion to make it easier to use the cuDNN batch normalization implementation. I'm creating an issue instead of a pull request to check if this would be a good idea, before I try to implement it.

As you can see in Keras (fchollet/keras#4162), you need quite a bit of code to properly use the cuDNN BN functions in the two gpu backends and have a fallback for when neither of those are available.

What would you think of adding abstract batch normalization ops to Theano? That is, two abstract ops that represent batch_normalization_train and batch_normalization_test, following the cuDNN semantics. Those ops can then be optimized into either one of the two gpu/cuDNN ops, or be replaced with an implementation using normal Theano ops for when cuDNN is not available.

Such functions would move much of this slightly awkward logic to Theano. For non-Keras purposes, this would also make it a lot easier to use these cuDNN BN functions, because you no longer have to write your code for a specific cpu/gpu backend.

The current `theano.tensor.nnet.bn.batch_normalization` is a bit different from the cuDNN functions, so I'd propose two new functions, for example, `theano.tensor.nnet.bn.batch_normalization_train` and `theano.tensor.nnet.bn.batch_normalization_test`, that follow exactly the cuDNN form.

**Comment by nouiz** [comment]
I like the idea. I wanted to have something like this in Theano since we
wrapped cudnn bn functions.

For the interfaces, I don't like the fact that the train and test function
have different interface. So I'm not sure we want to follow it directly add
the user interface. I would prefer an interface where both functions
signature are more similar. What do you think?

**Comment by f0k** [comment]
> For the interfaces, I don't like the fact that the train and test function have different interface.

Blame NVIDIA. `dnn_batch_normalization_train` returns `stdinv` because that's what cuDNN returns and requires for the gradient, and `dnn_batch_normalization_test` requires `var` instead because that's what cuDNN requires for inference.
The training Op could be extended to additionally perform the exponential moving averages, which are based on `var` and not `stdinv`. This requires inplace-versions along with an optimization, so one can do:
```
normed, mean, stdinv, new_running_mean, new_running_var = dnn_batch_normalization_train(x, beta, gamma, alpha, running_mean, running_var)
updates[running_mean] = new running_mean
updates[running_var] = new_running_var
```
and have this optimized to do in-place updates of running_mean and running_var. In addition, we may want to support in-place operation on the input, so the Op needs two separate in-place flags. And for performance we should still support the current operation without updating the exponential moving averages.

This being said, the abstract Op could support the full functionality (including moving averages) and implement it in Theano until the cuDNN Ops get extended.

**Comment by gvtulder** [comment]
I created a pull request that can serve as a starting point.

**Comment by lamblin** [comment]
Since the PR (and a follow-up) has been merged, can we close this?

**Comment by gvtulder** [comment]
Yes.
